app/api/domains/osu.py

```diff
@@ -617,6 +622,54 @@ async def osuSubmitModularSelector(
     score.bmap = bmap
     score.player = player
 
+    ## perform checksum validation
+
+    unique_id1, unique_id2 = unique_ids.split("|", maxsplit=1)
+    unique_id1_md5 = hashlib.md5(unique_id1.encode()).hexdigest()
+    unique_id2_md5 = hashlib.md5(unique_id2.encode()).hexdigest()
+
+    client_hash_decoded = aes.decrypt(b64decode(client_hash_b64)).decode()
+
+    try:
+        # TODO: assert osu_version
+        # TODO: assert client_hash
+        assert player.client_details is not None
+
+        # assert unique ids (c1) are correct and match login params
+        assert (
+            unique_id1_md5 == player.client_details.uninstall_md5
+        ), f"unique_id1 mismatch ({unique_id1_md5} != {player.client_details.uninstall_md5})"
+        assert (
+            unique_id2_md5 == player.client_details.disk_signature_md5
+        ), f"unique_id2 mismatch ({unique_id2_md5} != {player.client_details.disk_signature_md5})"
+
+        # assert online checksums match
+        server_score_checksum = score.compute_online_checksum(
+            osu_version=osu_version,
+            osu_client_hash=client_hash_decoded,
+            storyboard_checksum=storyboard_md5 or "",
+        )
+        assert (
+            score.client_checksum == server_score_checksum
+        ), f"online score checksum mismatch ({server_score_checksum} != {score.client_checksum})"
+
+        # assert beatmap hashes match
+        assert (
+            updated_beatmap_hash == bmap_md5
+        ), f"beatmap md5 checksum mismatch ({updated_beatmap_hash} != {bmap_md5}"
```

Comment:
these assertions can likely be cleaned up or simplified - not sure how to go about it yet